stellargraph/datasets/dataset_loader.py

```diff
@@ -0,0 +1,218 @@
+            )
+        )
+
+    def _resolve_path(self, filename: str) -> str:
+        """Convert dataset relative file names to their full path on filesystem"""
+        return os.path.join(self.base_directory, filename)
+
+    def _is_downloaded(self) -> bool:
+        """Returns true if the expected files for the dataset are present"""
+        return all(
+            os.path.isfile(self._resolve_path(file)) for file in self.expected_files
+        )
+
+    def _verify_files_downloaded(self) -> None:
+        """
+        Raises:
+            FileNotFoundError: If any files within dataset are missing.
+        """
+        missing_files = ", ".join(
+            [
+                file
+                for file in self.expected_files
+                if not os.path.isfile(self._resolve_path(file))
+            ]
+        )
+        if missing_files:
+            raise FileNotFoundError(
+                f"{self.name} dataset failed to download file(s): {missing_files} to {self.data_directory}"
+            )
+
+    def _delete_existing_files(self) -> None:
+        """ Delete the files for this dataset if they already exist """
+        for file in self.expected_files:
+            try:
+                os.remove(self._resolve_path(file))
+            except OSError:
+                pass
+
+    def download(self, ignore_cache: Optional[bool] = False) -> None:
+        """
+        Download the dataset (if not already downloaded)
+
+        Args:
+            ignore_cache bool, optional (default=False): Ignore a cached dataset and force a re-download.
+
+        Raises:
+            FileNotFoundError: If the dataset is not successfully downloaded.
+        """
+        if ignore_cache:
+            self._delete_existing_files()  # remove any existing dataset files to ensure we re-download
+
+        if ignore_cache or not self._is_downloaded():
+            log.info(
+                "%s dataset downloading to %s from %s",
+                self.name,
+                self.base_directory,
+                self.url,
+            )
+            filename, _ = urlretrieve(
+                self.url
+            )  # this will download to a temporary location
+            if self.url_archive_format is None:
+                # not an archive, so the downloaded file is our data and just needs to be put into place
+                self._create_base_directory()
+                move(filename, self._resolve_path(self.expected_files[0]))
+            else:
+                # archive, so it needs to be unpacked into the right directory
+                unpack_archive(
+                    filename, self._all_datasets_directory(), self.url_archive_format
+                )
+            self._verify_files_downloaded()
+        else:
+            log.info("%s dataset is already downloaded", self.name)
+
+
+def download_all_datasets(ignore_cache: Optional[bool] = False) -> bool:
```

Comment:
If this function is only for testing, maybe we could not even have it here? That is, just have the `for cls in DatasetLoader.__subclasses__()` loop in the tests?

Reply:
I've now replaced this with @pytest.mark.parametrize in the tests to effectively create a test for downloading each dataset.